tests/run/cpp_iterators.pyx

```diff
@@ -98,6 +103,34 @@ def test_custom():
     finally:
         del iter
 
+def test_custom2():
+    """
+    (deref(iter) instead of iter[0])
+    >>> test_custom()
+    [1.0, 2.0, 3.0]
+    """
+    cdef double* values = [1, 2, 3]
+    cdef DoublePointerIter* iter
+    try:
+        iter = new DoublePointerIter(values, 3)
+        return [x for x in deref(iter)]
+    finally:
+        del iter
+
+def test_custom_genexp():
+    """
+    >>> list(test_custom())
+    [1.0, 2.0, 3.0]
+    """
+    cdef double* values = [1, 2, 3]
+    cdef DoublePointerIterCanMakeEmpty* iter
+    try:
+        iter = new DoublePointerIterCanMakeEmpty(values, 3)
+        # TODO: Needs to copy once - currently copies twice
+        return (x for x in iter[0])
```

Comment:
To go into a generator expression `iter` now needs to be copyable and default initializable. Copyable makes sense (since it needs to outlive the destruction in finally), but default initializable is just because the scope class is created empty then filled up later. I don't think this will be easy to change. It also gets copied twice which isn't ideal.

This restrictions don't apply for comprehensions.